romana/kubernetes/listener.go

```diff
@@ -0,0 +1,364 @@
+func (l *kubeListener) Routes() common.Routes {
+	routes := common.Routes{}
+	return routes
+}
+
+// Name implements method of Service interface.
+func (l *kubeListener) Name() string {
+	return "kubernetes-listener"
+}
+
+// SetConfig implements SetConfig function of the Service interface.
+func (l *kubeListener) SetConfig(config common.ServiceConfig) error {
+	m := config.ServiceSpecific
+	l.kubeUrl = m["kubernetes_url"].(string)
+	if l.kubeUrl == "" {
+		return errors.New("kubernetes_url required.")
+	}
+	l.urlPrefix = m["url_prefix"].(string)
+	if l.urlPrefix == "" {
+		return errors.New("url_prefix required.")
+	}
+	l.segmentLabelName = m["segment_label_name"].(string)
+	if l.segmentLabelName == "" {
+		return errors.New("segment_label_name required.")
+	}
+
+	return nil
+
+}
+
+// Run configures and runs listener service.
+func Run(rootServiceURL string, cred *common.Credential) (*common.RestServiceInfo, error) {
+	clientConfig := common.GetDefaultRestClientConfig(rootServiceURL)
+	clientConfig.Credential = cred
+	client, err := common.NewRestClient(clientConfig)
+	if err != nil {
+		return nil, err
+	}
+	kubeListener := &kubeListener{}
+	config, err := client.GetServiceConfig(kubeListener)
+	if err != nil {
+		return nil, err
+	}
+	return common.InitializeService(kubeListener, *config)
+}
+
+// translateNetworkPolicy translates a Kubernetes policy into
+// Romana policy (see common.Policy) with the following rules:
+// 1. Kubernetes Namespace corresponds to Romana Tenant
+// 2. If Romana Tenant does not exist it is an error (a tenant should
+//    automatically be created when a namespace is added, however)
+func (l *kubeListener) translateNetworkPolicy(kubePolicy *KubeObject) (common.Policy, error) {
+	romanaPolicy := &common.Policy{Direction: common.PolicyDirectionIngress}
+	ns := kubePolicy.Metadata.Namespace
+	tenantUrl, err := l.restClient.GetServiceUrl("tenant")
+	if err != nil {
+		return *romanaPolicy, err
+	}
+	tenants := []tenant.Tenant{}
+	err = l.restClient.Get(fmt.Sprintf("%s/tenants/%s", tenantUrl, ns), tenants)
+	if err != nil {
+		return *romanaPolicy, err
+	}
+	if len(tenants) == 0 {
+		return *romanaPolicy, common.NewError("No tenant found under %s", ns)
+	}
+	if len(tenants) > 1 {
+		return *romanaPolicy, common.NewError("More than one tenant found under %s", ns)
+	}
+	tenantId := tenants[0].ID
+	tenantExternalId := tenants[0].ExternalID
+	
+	segmentName := kubePolicy.Spec.PodSelector[l.segmentLabelName]
+	if segmentName == "" {
+		return *romanaPolicy, common.NewError("Expected segment to be specified in podSelector part as '%s'", l.segmentLabelName)
+	}
+
+	segments := []tenant.Segment{}
+	err = l.restClient.Get(fmt.Sprintf("%s/tenants/%s/segments/%s", tenantUrl, ns, segmentName), segments)
+	if err != nil {
+		switch err := err.(type) {
+		case common.HttpError:
+			if err.StatusCode == 404 {
+				// Not found, so let's create a segment.
+				segreq := tenant.Segment{Name: segmentName}
```

Comment:
Yes, I agree. I found the other place where the tenants are created after I wrote this comment. It's ok. Maybe leave a note here as a comment about that.